upb/decode.c

```diff
@@ -608,6 +609,25 @@ static bool decode_tryfastdispatch(upb_decstate *d, const char **ptr,
   return false;
 }
 
+static const char *decode_mset_item(upb_decstate *d, const char *ptr,
+                                    upb_msg *msg, const upb_msglayout *layout) {
+  // We create a temporary upb_msglayout here and abuse its fields as temporary
+  // storage, to avoid creating lots of MessageSet-specific parsing code-paths:
+  //   1. We store 'layout' in item_layout.subs.  We will need this later as
+  //      a key to look up extensions for this MessageSet.
+  //   2. We use item_layout.fields as temporary storage to store the extension we
+  //      found when parsing the type id.
+  upb_msglayout item_layout = {
+      .subs = (const upb_msglayout_sub[]){{.submsg = layout}},
+      .fields = NULL,
+      .size = 0,
+      .field_count = 0,
+      .ext = _UPB_MSGEXT_MSET_ITEM,
+      .dense_below = 0,
+      .table_mask = -1};
+  return decode_group(d, ptr, msg, &item_layout, 1);
+}
+
 UPB_NOINLINE
 static const char *decode_msg(upb_decstate *d, const char *ptr, upb_msg *msg,
                               const upb_msglayout *layout) {
```

Comment:
at 150 lines, this method has gotten more than a bit unweildy.  Can extract some of the parts of it to well named helper functions?

ordinarily, I would say a follow up CL is fine, but actually I am having a lot of trouble piecing this together to review, so it might be better to do the cleanup in an separate preceding CL and then rebase this onto that

Reply:
Done.